### PR TITLE
MUD

### PR DESCRIPTION
**Pull request by isstuev**
## Description and Related Issue(s)

Added `NEXT_PUBLIC_HAS_MUD_FRAMEWORK` variable to enable MUD-framework feature.

## Checklist for PR author
- [ ] I have tested these changes locally.
- [ ] I added tests to cover any new functionality, following this [guide](./CONTRIBUTING.md#writing--running-tests)
- [ ] Whenever I fix a bug, I include a regression test to ensure that the bug does not reappear silently.
- [ ] If I have added, changed, renamed, or removed an environment variable, I have updated the list of environment variables in the [documentation](ENVS.md) and  made the necessary changes to the validator script according to the [guide](./CONTRIBUTING.md#adding-new-env-variable)

### CODE REVIEW

File: ui/address/mud/AddressMudRecordsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  const onRecordClick = React.useCallback((e: React.MouseEvent) => {
+    if (e.metaKey || e.ctrlKey) {
+      // Allow opening in a new tab/window with right-click or ctrl/cmd+click
+      return;
+    }
+
+    e.preventDefault();
+
+    router.push(
+      { pathname: '/address/[hash]', query: { hash, tab: 'mud', table_id: data.table.table_id, record_id: e.currentTarget.getAttribute('data-id') as string } },
```
Comment (tom2drum) [inline ui/address/mud/AddressMudRecordsTable.tsx:74]: I would have checked whether the attribute value existed before navigating the page.

File: ui/address/mud/AddressMudRecordsTable.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    (e: React.MouseEvent) => toggleSorting('key' + e.currentTarget.getAttribute('data-id') as AddressMudRecordsSorting['sort']),
+    [ toggleSorting ],
+  );
+
+  const keyBgColor = useColorModeValue('blackAlpha.50', 'whiteAlpha.50');
+
+  React.useEffect(() => {
+    if (hasCut && !colsCutCount && containerRef.current) {
+      const count = Math.floor((containerRef.current.getBoundingClientRect().width - CUT_COL_WIDTH) / COL_MIN_WIDTH);
+      if (totalColsCut > 2 && count - 1 < totalColsCut) {
```
Comment (tom2drum) [inline ui/address/mud/AddressMudRecordsTable.tsx:99]:
```suggestion
      if (totalColsCut > MIN_CUT_COUNT && count - 1 < totalColsCut) {
```

🙄

File: ui/shared/sort/getNextSortValue.ts [hunk 3]
```diff
@@ -3,8 +3,20 @@ export default function getNextSortValue<SortField extends string, Sort extends
 ) {
   return (prevValue: Sort | undefined) => {
     const sequence = sortSequence[field];
+    getNextValueFromSequence(sequence, prevValue);
     const curIndex = sequence.findIndex((sort) => sort === prevValue);
     const nextIndex = curIndex + 1 > sequence.length - 1 ? 0 : curIndex + 1;
     return sequence[nextIndex];
```
Comment (tom2drum) [inline ui/shared/sort/getNextSortValue.ts:9]: Should we delete these lines then?

File: ui/shared/filters/TableColumnFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  title: string;
+  isActive?: boolean;
+  isFilled?: boolean;
+  onFilter: () => void;
+  onReset?: () => void;
+  onClose?: () => void;
+  isLoading?: boolean;
+  className?: string;
+  children: React.ReactNode;
+}
```
Comment (tom2drum) [inline ui/shared/filters/TableColumnFilter.tsx:23]: Maybe it is better to inherit some fields from `ContentProps`

File: ui/shared/DataListDisplay.tsx [hunk 5]
```diff
@@ -13,8 +13,9 @@ type FilterProps = {
 type Props = {
   isError: boolean;
   items?: Array<unknown>;
-  emptyText: string;
+  emptyText: string | React.ReactNode;
```
Comment (tom2drum) [inline ui/shared/DataListDisplay.tsx:16]:
```suggestion
  emptyText: React.ReactNode;
```

`string` is also a `ReactNode`

File: ui/pages/Address.tsx [hunk 6]
```diff
@@ -98,7 +107,10 @@ const AddressPageContent = () => {
     undefined;
 
   const isLoading = addressQuery.isPlaceholderData || (config.features.userOps.isEnabled && userOpsAccountQuery.isPlaceholderData);
```
Comment (tom2drum) [inline ui/pages/Address.tsx:109]: Should the condition about UserOps be moved into `isTabsLoading` too?

File: ui/mudWorlds/MudWorldsTableItem.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  return (
+    <Tr>
+      <Td verticalAlign="middle">
+        <AddressEntity address={ item.address } isLoading={ isLoading } fontWeight={ 700 }/>
+      </Td>
+      <Td isNumeric>
+        <Skeleton isLoaded={ !isLoading } display="inline-block" maxW="100%">
+          <Text lineHeight="24px" as="span">{ addressBalanceChunks[0] }</Text>
+          { addressBalanceChunks[1] && <Text lineHeight="24px" as="span">.</Text> }
```
Comment (tom2drum) [inline ui/mudWorlds/MudWorldsTableItem.tsx:24]: This "." could be moved inside the previous element.

File: ui/address/mud/AddressMudTablesTableItem.tsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+  const onTableClick = React.useCallback((e: React.MouseEvent) => {
+    if (e.metaKey || e.ctrlKey) {
+      // Allow opening in a new tab/window with right-click or ctrl/cmd+click
+      return;
+    }
+
+    e.preventDefault();
+
+    router.push(
```
Comment (tom2drum) [inline ui/address/mud/AddressMudTablesTableItem.tsx:33]: Same comment

File: ui/address/mud/AddressMudTablesListItem.tsx [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+  const onTableClick = React.useCallback((e: React.MouseEvent) => {
+    if (e.metaKey || e.ctrlKey) {
+      // Allow opening in a new tab/window with right-click or ctrl/cmd+click
+      return;
+    }
+
+    e.preventDefault();
+
+    router.push(
```
Comment (tom2drum) [inline ui/address/mud/AddressMudTablesListItem.tsx:35]: 🤖

File: ui/address/mud/AddressMudTable.tsx [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    setFilters(prev => {
+      const newFilters = { ...prev };
+      delete newFilters[key];
+      return newFilters;
+    });
+  }, []);
+
+  const hasActiveFilters = Object.values(filters).some(Boolean);
+
+  const actionBatHeight = React.useMemo(() => {
```
Comment (tom2drum) [inline ui/address/mud/AddressMudTable.tsx:68]: What is "Bat"? 🦇

File: ui/address/mud/AddressMudTable.tsx [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    const heightWithoutFilters = pagination.isVisible ? ACTION_BAR_HEIGHT_DESKTOP : 60;
+
+    return hasActiveFilters ? heightWithoutFilters + 44 : heightWithoutFilters;
```
Comment (tom2drum) [inline ui/address/mud/AddressMudTable.tsx:71]: No magic numbers, pls 🔮

File: ui/shared/sort/getNextSortOrder.tsx
Comment (tom2drum) [inline ui/shared/sort/getNextSortOrder.tsx:1]: 🤔

**Review by tom2drum — APPROVED** [review]

File: lib/capitalizeFirstLetter.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+export default function capitalizeFirstLetter(text: string) {
```
[the remark this answers is not in the document]
Comment (isstuev) [reply lib/capitalizeFirstLetter.ts:1]: _capitalize work like this: qWeRtY -> Qwerty, but I need qWeRtY -> QWeRtY

File: ui/pages/Address.tsx (hunk 6 above)
Comment (isstuev) [reply ui/pages/Address.tsx:109]: reasonable

**Pull request merged**